File: util/test_gj_solve.py
#!/usr/bin/env python
"""
Copyright (c) 2016, Donald E. Willcox
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of gauss-jordan-solver nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""
import numpy as np
import os
import argparse
from SparseGaussJordan import GaussJordan, Row

def csr_matrix_vector_dot(GJ, A, x):
    b = np.zeros_like(x)
    
    icsr = 0
    for irow, r in enumerate(GJ.sparsity):
        for icol in range(GJ.nvars):
            if not r.zero_at(icol):
                b[irow] += A[icsr] * x[icol]
                icsr += 1
    return b

def random_sample(GJ, compressed_sparse_row, scale, verbose=False):
    x = scale * np.random.rand(GJ.nvars)
    if compressed_sparse_row:
        A = np.zeros(GJ.get_number_nonzero(), dtype=np.float64)
    else:
        A = np.zeros((GJ.nvars, GJ.nvars), dtype=np.float64)
    icsr = 0
    for j, r in enumerate(GJ.sparsity):
        for k in range(GJ.nvars):
            if not r.zero_at(k):
                frand = np.random.rand()
                while frand==0.0:
                    # Get a nonzero random number for element of A.
                    frand = np.random.rand()
                if compressed_sparse_row:
                    A[icsr] = scale * frand
                else:
                    A[j][k] = scale * frand
                if verbose:
                    print('A[{}][{}] = random'.format(j, k))
                icsr += 1
            else:
                if not compressed_sparse_row:
                    A[j][k] = 0.0
                if verbose:
                    print('A[{}][{}] = 0'.format(j, k))
    return x, A

def test_gj_solve(GJ, compressed_sparse_row, ntest, scale, tol):
    # Test the Gauss-Jordan solver by randomizing matrix elements and verifying output
    xresd_accumulate = np.empty((ntest, GJ.nvars), dtype=np.float64)
    xtrue, A = random_sample(GJ, compressed_sparse_row, scale, verbose=True) # Print which elements get set
    print(A)
    for i in range(ntest):
        xtrue, A = random_sample(GJ, compressed_sparse_row, scale)
        if compressed_sparse_row:
            b = csr_matrix_vector_dot(GJ, A, xtrue)
        else:
            b = np.dot(A, xtrue)
        xtest = gauss_jordan_solve(A, b)
        xresd = xtrue-xtest # Absolute value of difference
        xresd_accumulate[i][:] = xresd[:]
    # Get statistics
    xresd_ave = np.average(xresd_accumulate, axis=0)
    xresd_std = np.std(xresd_accumulate, axis=0)
    # Print statistics
    print('----------------------------------------')
    print('Residual Average +/- StdDev for solutions of the system Ax=b.')
    print('Using {} random iterations.'.format(ntest))
    is_okay = True
    for i, xstat in enumerate(zip(xresd_ave, xresd_std)):
        print('Residual for X[{}]: {} +/- {}'.format(i, xstat[0], xstat[1]))
        if xstat[0] > tol or xstat[1] > tol:
            is_okay = False
            print('Tolerance ERROR for X[{}]!'.format(i))
            print('Tolerance is {}'.format(tol))
    if is_okay:
        print('System OK!\nAverage and standard deviation of all residuals is below the tolerance {}'.format(tol))
        
if __name__=='__main__':
    ### MAIN ###
    parser = argparse.ArgumentParser()
    parser.add_argument('structurefile', type=str,
                        help="""Name of the file specifying sparsity pattern 
                        of the matrix on which to test the gauss-jordan solution.""")
    parser.add_argument('-n', type=int, default=15,
                        help='Number of random matrices to test. (Default is 15).')
    parser.add_argument('-scale', type=float, default=100.0,
                        help='Scaling factor with which to multiply the random matrix elements.')
    parser.add_argument('-tol', type=float, default=1.0e-12,
                        help="""Round-off error tolerance for residuals (both average and stdev) 
                        below which the solution is OK. (Default is 1.0E-12).""")
    parser.add_argument('-csr', action='store_true',
                        help='Use compressed sparse row (CSR) matrix format.')
    parser.add_argument('-smp', action='store_true',
                        help="""Attempt to simplify solution. Can be very slow, but if possible, 
                        will reduce the number of operations required for the solution.""")
    parser.add_argument('-expand', action='store_true',
                        help="""Simplify solution by expanding it and performing cancellations.""")    
    parser.add_argument('-cse', action='store_true',
                        help="""Execute Common Subexpression Elimination. 
                        (After simplification if the -smp option is present.) 
                        This may be more helpful for Python than a compiled language 
                        with a compiler capable of weighing the operation and memory costs.""")
    parser.add_argument('-keep', action='store_true',
                        help="""Do not delete 'test.py' output solver once the test is complete.""")
    args = parser.parse_args()

    # Generate the python script file 'test.py' to solve the system
    sfile = args.structurefile
    GJ = GaussJordan(sfile, args.csr, 'test.py', None, args.smp, args.expand, args.cse)
    from test import gauss_jordan_solve
    test_gj_solve(GJ, args.csr, args.n, args.scale, args.tol)

    # Delete the 'test.py' file if user didn't decide to keep it
    if not args.keep:
        here = os.getcwd()
        os.remove(os.path.join(here, 'test.py'))

File: util/test_gj_solve.f90
! Copyright (c) 2016, Donald E. Willcox
! All rights reserved.
!
! Redistribution and use in source and binary forms, with or without
! modification, are permitted provided that the following conditions are met:
!
! * Redistributions of source code must retain the above copyright notice, this
!   list of conditions and the following disclaimer.
!
! * Redistributions in binary form must reproduce the above copyright notice,
!   this list of conditions and the following disclaimer in the documentation
!   and/or other materials provided with the distribution.
!
! * Neither the name of gauss-jordan-solver nor the names of its
!   contributors may be used to endorse or promote products derived from
!   this software without specific prior written permission.
!
! THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
! AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
! IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
! DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
! FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


subroutine get_nonzero_dp(x)
  implicit none
  
  double precision :: x
  x = 0.0d0
  do while ( x == 0.0d0)
     call RANDOM_NUMBER(x)
  end do
end subroutine get_nonzero_dp

subroutine random_solve(which_iters, which_solver, N, niter, Asparse, aresid)
  use gauss_jordan_module, only : gauss_jordan_solve
  implicit none

  integer, intent(in) :: N, which_solver, which_iters, niter
  integer, dimension(N,N), intent(in) :: Asparse
  double precision, dimension(N), intent(out) :: aresid
  double precision, dimension(N,N) :: A, Acopy
  double precision, dimension(N) :: xsolv, xtrue, b
  integer, dimension(N) :: IPIV
  integer :: i, j, INFO

  do i = 1, N
     call get_nonzero_dp(xtrue(i))
     do j = 1, N
        if ( Asparse(i, j) == 1 ) then
           call get_nonzero_dp(A(i, j))
        else
           A(i, j) = 0.0d0
        end if
     end do
  end do
  b = MATMUL(A, xtrue)
  if ( which_solver == 1 ) then
     !     write(*,*) 'Using gauss-jordan solver'
     if ( which_iters == 2 ) then
        do i = 1, niter
           Acopy = A
           call gauss_jordan_solve(Acopy, xsolv, b)
        end do
     elseif ( which_iters == 1 ) then
        call gauss_jordan_solve(A, xsolv, b)
     end if
  elseif ( which_solver == 2) then
     !     write(*,*) 'Using LAPACK DGESV'
     if ( which_iters == 2 ) then
        do i = 1, niter
           xsolv = b
           Acopy = A
           call DGESV(N, 1, Acopy, N, IPIV, xsolv, N, INFO)
        end do
     elseif ( which_iters == 1 ) then
        xsolv = b
        call DGESV(N, 1, A, N, IPIV, xsolv, N, INFO)
     end if
     if ( INFO /= 0 ) then
        write(*,*) 'LAPACK could not find a solution!'
        stop
     end if
  else
     write(*,*) 'NO SOLVER SPECIFIED. ENTER 1 or 2.'
     stop
  end if
  aresid = abs(xtrue - xsolv)
end subroutine random_solve

program test_gj_solve
  ! Takes 4 command-line arguments:
  ! 1: Name of sparsity structure file to read
  ! 2: Number of random matrix equations Ax=b to solve
  ! 3: Which solver to use (1 = gauss jordan, 2 = LAPACK DGESV)
  ! 4: Matrix generation (1 = resample matrix every iteration,
  !                       2 = randomly sample matrix only once)
  
  implicit none
  
  character(LEN=50) :: sparsity_file
  character(LEN=50) :: str_num_iters, str_solver_choice, str_iters_choice
  integer :: N, unit, niter, which_solver, which_iters
  integer, dimension(:,:), allocatable :: Asparsity
  double precision, dimension(:,:), allocatable :: aresid
  double precision, dimension(:), allocatable :: arave, arstd
  double precision :: scratch
  integer :: i, j
  logical :: invalid_arguments

  ! Initialize the random number generator
  call init_random_seed()

  invalid_arguments = .false.
  
  ! Read the first command-line argument as the sparsity file name
  call GET_COMMAND_ARGUMENT(1, sparsity_file)

  if (len(trim(sparsity_file)) .eq. 0) then
     invalid_arguments = .true.
  end if

  ! Read the second command-line argument as the number of random solutions to run
  call GET_COMMAND_ARGUMENT(2, str_num_iters)
  read(str_num_iters, '(I50)') niter
  if ( niter <= 0 ) then
     invalid_arguments = .true.
  end if
  ! Read the third command-line argument as the selection for which solver to use
  ! if which_solver = 1, then use the nonzero gauss-jordan solver
  ! if which_solver = 2, then use LAPACK DGESV
  call GET_COMMAND_ARGUMENT(3, str_solver_choice)
  read(str_solver_choice, '(I50)') which_solver
  if ( .not. ( which_solver == 1 .or. which_solver == 2 )) then
     invalid_arguments = .true.
  end if
  if ( which_solver == 1 ) then
     write(*,*) 'Using Nonzero Gauss-Jordan Solver'
  elseif ( which_solver == 2 ) then
     write(*,*) 'Using LAPACK DGESV Solver'
  end if
  ! Read the fourth command-line argument as the selection for which randomization to use
  ! if which_iters = 1, then solve different randomized matrix systems
  ! if which_iters = 2, then solve the same randomized matrix system over and over
  call GET_COMMAND_ARGUMENT(4, str_iters_choice)
  read(str_iters_choice, '(I50)') which_iters
  if ( .not. ( which_iters == 1 .or. which_iters == 2 )) then
     invalid_arguments = .true.
  end if

  if (invalid_arguments) then
     print *, "test_gj_solve takes 4 command-line arguments:"
     print *, "  1: Name of sparsity structure file to read"
     print *, "  2: Number of iterations of matrix equations Ax=b to solve"
     print *, "  3: Which solver to use (1 = gauss jordan, 2 = LAPACK DGESV)"
     print *, "  4: Matrix generation (1 = resample random matrix every iteration,"
     print *, "                        2 = randomly sample matrix only once)"
     stop
  end if

  ! Read the sparsity file and set up sparsity matrix
  open(newunit=unit, file=sparsity_file)
  ! Get size of matrix
  read(unit,*) N
  allocate( Asparsity(N, N) )
  ! Read sparsity matrix in row-major order
  read(unit,*) Asparsity
  ! Transpose matrix into column-major format for Fortran
  Asparsity = TRANSPOSE(Asparsity)
  close(unit=unit)
  
  ! Test solver with a random matrix equation using this sparsity
  allocate( arave(N) )
  allocate( arstd(N) )
  if ( which_iters == 1 ) then
     allocate( aresid(N, niter) )
     do i = 1, niter
        call random_solve(which_iters, which_solver, N, niter, Asparsity, aresid(:, i))
     end do
     ! Get statistics for residuals
     do i = 1, N
        arave(i) = SUM(aresid(i, :))/dble(niter)
        scratch = 0.0d0
        do j = 1, niter
           scratch = scratch + ((aresid(i, j) - arave(i))**2)/dble(niter)
        end do
        arstd(i) = sqrt(scratch)
     end do
     ! Print statistics for absolute values of residuals
     write(*,*) niter, ' Random Systems Solved'
     write(*,*) 'Residual Average +/- StdDev for solutions of the system Ax=b.'
     do i = 1, N
        write(*,*) 'x(', i, '): ', arave(i), ' +/- ', arstd(i)
     end do
  elseif ( which_iters == 2 ) then
     allocate( aresid(N, 1) )
     call random_solve(which_iters, which_solver, N, niter, Asparsity, aresid(:, 1))
     ! Print residuals
     write(*,*) niter, ' Solution Iterations of a Random System'
     write(*,*) 'Residuals for solution of the system Ax=b.'
     do i = 1, N
        write(*,*) 'x(', i, '): ', aresid(i, 1)
     end do
  end if

  ! Cleanup memory
  deallocate( aresid )
  deallocate( arave )
  deallocate( arstd )
  deallocate( Asparsity )
end program test_gj_solve
